### ISSUE TITLE
Required warning no emitted on update to 2024 edition

### ISSUE THREAD

**Issue by weiznich** [rust-lang/rust#133541]
Steps to reproduce:

```sh
# 1.56.0 corresponds to the first rust version before of the 2021 edition
cargo +1.56.0 new test_edition --lib 
cd test_edition
cat << EOF > Cargo.toml
[package]
name = "test_edition"
version = "0.1.0"
edition = "2021"
rust-version = "1.65.0"

# just using diesel as example here
# as I know the crate well
# the issue is far more general than about diesel
[dependencies]
diesel = { version = "2.0", features = ["sqlite"] }
EOF


cat << EOF > src/lib.rs           
use diesel::connection::InstrumentationEvent;
use diesel::prelude::*;

pub fn foo() {
    let mut conn = SqliteConnection::establish(":memory:").unwrap();

    // use an api from diesel 2.2
    conn.set_instrumentation(|event: InstrumentationEvent<'_>| println!("{:?}", event));

    println!("Hello, world!");
}
EOF

git add .
git commit -m "First commit"

# simulate a new clone
git clean -Xfd .

# compiles without warning
cargo +nightly check

# bump edition to 2024, which should be fine
cat << EOF > Cargo.toml
[package]
name = "test_edition"
version = "0.1.0"
# do not actually bump the edition yet
# as there are no packages yet with a minimal supported rust version at the edition boundary
# that contain new features
# It shouldn't be too hard to see that this will automatically happen as soon as the edition passed
# by for some month
edition = "2021"
# use resolver = 3 to simulate this behavior
# just with lower rust versions
# (This will be the new default in edition 2024)
resolver = "3"
rust-version = "1.65.0"

# just using diesel as example here
# as I know the crate well
# the issue is far more general than about diesel
[dependencies]
diesel = { version = "2.0", features = ["sqlite"] }
EOF

git add Cargo.toml
git commit -m "Bump edition"

# simulate a new clone
git clean -Xfd .

cargo +nightly check
# observe a build failure
```

The expected behavior here is that cargo emits a warning about the imprecise dependency before we update the edition. [RFC-2052](https://github.com/rust-lang/rfcs/blob/master/text/2052-epochs.md#hard-constraints) literally writes this as hard constraint for any edition change:

> TL;DR: Warning-free code on edition N must compile on edition N+1 and have the same behavior.

I put this here again as I already raised this several times before with the cargo team, but I did not get any proper response to this up until today. I only got responses along the line that the dependency declaration is already broken and that's not an issue. This is demonstrably not true yet, as you can build the corresponding package with the 2021 edition/resolver 2 behavior without warnings, so this behavior is not considered yet to be problematic. I agree in so far that you could produce the same error by manually downgrading dependencies, but that's not the default behavior. 

### Meta

`rustc --version --verbose`:
```
rustc +nightly --version --verbose 
rustc 1.85.0-nightly (dff3e7ccd 2024-11-26)
binary: rustc
commit-hash: dff3e7ccd4a18958c938136c4ccdc853fcc86194
commit-date: 2024-11-26
host: x86_64-unknown-linux-gnu
release: 1.85.0-nightly
LLVM version: 19.1.4
```

@rustbot label: +A-edition-2024

**Comment by fmease** [comment]
> I put this here again as I already raised this several times before with the cargo team, but I did not get any proper response to this up until today

I'm sorry to hear that. I guess it makes sense for gaining visibility. However, I think the Cargo issue tracker is still better suited for the technical side of things. Hmm, or Zulip. 

> > TL;DR: Warning-free code on edition N must compile on edition N+1 and have the same behavior.

Hmm, I never read the RFC before I'm surprised by this statement because it doesn't reflect reality.
In practice for most (all?) things we don't actually issue any edition-specific warnings by default.
You'd have bump the level of the respective `rustc_XXXX_compatibility` lint group from allow.

---

Still, there doesn't seem to be an equivalent [Cargo lint](https://doc.rust-lang.org/nightly/cargo/reference/lints.html) group (and Cargo lints themselves are apparently still unstable?), so in the end you seem to be right in stating that there's no edition warning / migration lint for `msrv-resolver` / `resolver = "3"`. I'm not an Cargo expert tho.

**Comment by jieyouxu** [comment]
FYI cc @traviscross for a potential edition concern.

**Comment by ehuss** [comment]
I'm going to close, as I do not believe this is the proper place to raise this grievance. The edition team also does not believe that a change is needed here. We have asked that if you have further concerns, that you bring it up with the moderation team at rust-mods@rust-lang.org, and we ask you to do that again if you have such concerns.
